### PR TITLE
Datasource Onboarding: Prevent flickering of onboarding page after first load

### LINKED ISSUE #63609: Datasource Onboarding: empty state flickers after creating the first datasource
**What happened**:

Showing the DS onboarding page is currently dependant on the `state.dataSources.hasFetched` state property, which once set to `true` after loading the list of datasource instances the first time, does never get set back to `false`. this causes a flicker (as in showing the onboarding page briefly) when navigating to a new dashboard or explore because - while `hasFetched` is still set to `true` - we need to refetch datasources every times that page renders to also decide wether to show that page in the first place or not:

https://user-images.githubusercontent.com/1170767/219000513-4a008e98-074c-4835-accf-83281e069dfa.mov

**What you expected to happen**:

empty state should not be briefly shown after adding the first datasource

**How to reproduce it (as minimally and precisely as possible)**:

with `datasourceOnboarding` feature toggle on, make sure you have no data source setup (is using devenv comment out everything in `devenv/datasources.yaml` and remove grafana db):
- go to new dashboard (or explore), the onboarding page should appear
- create a datasource
- try to create a new dashboard (or navigate to explore), the onboarding page is briefly shown.

**Anything else we need to know?**:

**Environment**:
- Grafana version: main

### PR DESCRIPTION
**Pull request by Elfo404**
Showing the DS onboarding page is currently dependant on the `state.dataSources.hasFetched` state property, which once set to `true` after loading the list of datasource instances the first time, does never get set back to `false`. this causes a flicker (as in showing the onboarding page briefly) when navigating to a new dashboard or explore because - while `hasFetched` is still set to `true` - we need to refetch datasources every times that page renders to also decide wether to show that page in the first place or not:

https://user-images.githubusercontent.com/1170767/219000513-4a008e98-074c-4835-accf-83281e069dfa.mov

(see #63609 for more details)

this PR fixes that by slightly refactoring the dasources state by renaming  `isLoadingDataSources` to `isLoadingDataSourcePlugins`, and use `isLoadingDataSources` to indicate if the datasource list is currently being loaded

Fixes #63609

**How to test this**:
with `datasourceOnboarding` feature toggle on, make sure you have no data source setup (is using devenv comment out everything in `devenv/datasources.yaml` and remove grafana db):

- go to new dashboard (or explore), the onboarding page should appear
- create a datasource
- try to create a new dashboard (or navigate to explore), the onboarding page should not briefly show up.

### CODE REVIEW

**Review by ifrost — APPROVED** [review]
LGTM! I think it's also worth extracting common logic to a separate component.

**Comment by Elfo404** [comment]
@ifrost this was still a draft as i think there are better ways to handle this and was just meant to be a 1 min poc to workaround the issue, i'm planning to properly fix this by modifying the state like so:
- remove `hasFetched` as it is currently used like a `loading` indicator, but it only represents wether datasource have been loaded at least once (as in it's set to true once and the never updated)
- add an `isLoading` item that should do what this PR does, but in a centralized way so no code repetition is needed.
- return `isLoading` alongside the list of datasources from `useLoadDataSources` and use that in both dashboards and explore.

i should have a PR ready shortly and i'll re-ask for review

File: public/app/features/datasources/state/reducers.ts [hunk 1]
```diff
@@ -17,13 +17,14 @@ export const initialState: DataSourcesState = {
...
   dataSourceTypeSearchQuery: '',
-  hasFetched: false,
   isLoadingDataSources: false,
+  isLoadingDataSourcePlugins: false,
   dataSourceMeta: {} as DataSourcePluginMeta,
   isSortAscending: true,
 };
 
 export const dataSourceLoaded = createAction<DataSourceSettings>('dataSources/dataSourceLoaded');
+export const dataSourcesLoad = createAction<void>('dataSources/dataSourcesLoad');
```
Comment (ifrost) [inline public/app/features/datasources/state/reducers.ts:27]: Naming is a bit confusing to me (maybe it's only me). Something like `dataSourcesLoadingStarted` would be a bit clearer but I see you're just using the pattern used in other actions so you don't need to change it (unless you also feel it's confusing and you want to change other names as well).

**Review by ifrost — COMMENTED** [review]
Looks nice! Flickering is gone when going back to Explore/Dashboards but I see a new flickering happening on the onboarding page. After selecting a data source it looks like the list refreshes and only then it opens up the settings. I don't see it happening in main.

Could you check if you can reproduce it? Videos for the reference:

main: 

https://user-images.githubusercontent.com/745532/220881722-fdf1de15-bdaa-43a2-a4fd-9cf0864f478c.mov 

branch:

https://user-images.githubusercontent.com/745532/220881788-e85432a3-6e25-45b1-a675-b3971bd3811b.mov

**Comment by Elfo404** [comment]
@ifrost good catch! i had to slow down network speed to actually se that happen. the reason behind it is that when adding a new datasource, the same action gets called, which triggers the loading state. i don't think there's any need for such state when adding a datasource so i changed it to reflect that in 8e209be7ba. wdyt?

File: public/app/features/datasources/state/reducers.ts (hunk 1 above)
Comment (Elfo404) [reply public/app/features/datasources/state/reducers.ts:27]: yeah, i was also kinda bothered by it, but didn't want to create inconsistencies within that file and i'd rather not mix refactors and actual changes, but happy to do it if you feel strongly about it.

**Review by ifrost — APPROVED** [review]
> (...) the reason behind it is that when adding a new datasource, the same action gets called, which triggers the loading state. i don't think there's any need for such state when adding a datasource so i changed it (...)

Works great 👍  I wonder why loading data sources is needed at all, as data sources should already be in the store in this case, right? Is there any other action using it before loading data sources? If yes, maybe a short comment in the code could help making it a bit clearer.

**Comment by Elfo404** [comment]
@ifrost i guess it tries to ensure the list is up to date before adding a new one.  the only way this can be practically useful is on the same function at line 215:

```ts
if (nameExits(dataSources, newInstance.name)) {
  newInstance.name = findNewName(dataSources, newInstance.name);
}
```
where we check whether a ds with the same name already exists and assign the new one a new name if needed. Will add a comment explaining this

**Pull request merged**